### ISSUE TITLE
[5.0] Incorrect initial value of $modules for onPrepareModuleList event

### ISSUE THREAD

**Issue by regularlabs** [joomla/joomla-cms#42149]
A bit late to the party. But there is still a bit of an issue in the `ModuleHelper::load` method (`libraries/src/Helper/ModuleHelper.php`).
In Joomla 4 it initially passes the `$modules` object to the `onPrepareModuleList` event as `null`.
This way plugins can see the difference between an unset or an empty module list.

In Joomla 5 however, it is initially set to an empty array. So plugins can no longer tell the difference.

That difference is important!

Say you have 2 plugins that do something on `onPrepareModuleList`.
Plugin 1 checks stuff and decides the module list should be empty (= no modules should be rendered).
Plugin 2 sees an empty array and says: "Ok, another plugin already decided there are no modules. So we don't have to do anything.

Or:
Plugin 1 checks stuff and decides it should not do anything. So it leaves the `$modules` as it is (`null`)
Plugin 2 sees the `null` value and says: "Ok, we are first to the party. Lets grab the modules we want to push into the `$modules` list.

Again: no longer possible in Joomla 5.
So now a second plugin would have to override the empty array module list even if a previous module purposely set it that way.

So can the `$modules = [];` be changed back to `$modules = null;` in the `ModuleHelper::load` method?

**Comment by Fedik** [comment]
It is late to do such change at this point. 
It was made to keep data consistent, to be sure it is always an array https://github.com/joomla/joomla-cms/blob/d38a8ffb0d8f2b4579422fc3d229eb60ffa1b1be/libraries/src/Event/Module/ModuleListEvent.php#L73

However, we can add extra argument to `PrepareModuleListEvent` class, that will flag if array is initialised or not.
In theory

**Comment by bembelimen** [comment]
This sounds like a very unreliable way of checking if modules are loaded or not.

Wouldn't it be better to communicate it by either asking the other plugin, storing the info somewhere or as Fedir suggested by adding more arguments?

**Comment by regularlabs** [comment]
I'm just stating how it was done pre-Joomla5.
With Joomla 5 the behaviour has changed. So a breaking change.
And for now, no way to see if the $modules array is empty because it is purposely made empty, or it hasn't been looked at yet.

I take it that any change to this will have to wait till J6 (and will probably be done completely different then).
So I'll just have to override the empty $modules array in the Joomla 5 version of my plugin. Which can cause conflicts with other plugins. Oh well...

**Comment by Fedik** [comment]
Please review and test https://github.com/joomla/joomla-cms/pull/42197

**Comment by regularlabs** [comment]
> This sounds like a very unreliable way of checking if modules are loaded or not.
Actually, as @Fedik noticed in the core code (seeing there is a fix there too in the PR), Joomla uses this method itself to see if any modules have purposely set the modules array to an empty array or if it is still false.

```
        // If the onPrepareModuleList event returns an array of modules, then ignore the default module list creation
        if (!$modules) {
            $modules = static::getModuleList();
        }
```

So setting the $modules to the empty array to begin with, breaks that core code in the same method.

So yeah, the added `setLoaded` in https://github.com/joomla/joomla-cms/pull/42197 is a different way to solve it.
But it completely changes the way this now works.

So in other words, the `$modules = [];` is not just a BC break, but also a bug.
And adding a `setLoaded` is a new feature and is not a fix for the bug.

**Comment by ssnobben** [comment]
So what did this go to?

**Comment by alikon** [comment]
the pr #42197 has been closed, so reopening this issue

**Comment by Fedik** [comment]
There no reason to keep this open, the `$module` will stay as array.

**Comment by alikon** [comment]
ok

**Comment by regularlabs** [comment]
@Fedik Ok, so you are saying you are keeping the initial `$modules` as an empty array (and noot `null`), forcefully creating a BC break. Ok, clear. I'll hack my way around things again.

Another case of changes for changes sake, without accounting for how things are used by extensions (the life-blood of Joomla).

**Comment by Fedik** [comment]
The event purpuse is to allow to initialise the modules list outside, by extensions. Look original PR https://github.com/joomla/joomla-cms/pull/6320
If you need to filter/add/remove modules, then there 2 more events for that:
`onAfterModuleList`
`onAfterCleanModuleList`


> Another case of changes for changes sake

It is for data consistency.

**Comment by regularlabs** [comment]
Here is another message you can mark as abusive without reason.
A shame nothing ever changes in the way people are treated by core developers.

![image](https://github.com/joomla/joomla-cms/assets/13553242/c3ff45fb-8e10-4d4c-9237-e20ce836a427)

**Comment by ssnobben** [comment]
Please core devs and team make a positive open environment for all contributing people on all levels that are not coding the Joomla core but important involved with the project long term. 

With all discussed and important changes its good to make all people understand similar arguments analysis like the Zachman framework analysis: why, how, what, who, where to get answered in a polite manner. 
![The_Zachman_Framework_of_Enterprise_Architecture](https://github.com/joomla/joomla-cms/assets/6490832/d2691dd3-af34-4bd5-96e2-3510edc9f0e9)

Thks.
